[Ready for Review] Translate Zero-to-Binder tutorial into Julia

sgibson91 opened the pull request:
<!--
Please complete the following sections when you submit your pull request. You are encouraged to keep this top level comment box updated as you develop and respond to reviews. Note that text within html comment tags will not be rendered.
-->
### Summary

<!-- Describe the problem you're trying to fix in this pull request. Please reference any related issue and use fixes/close to automatically close them, if pertinent. For example: "Fixes #58", or "Addresses (but does not close) #238". -->

Fixes #952 

This Pull Request translates the Zero to Binder tutorial into Julia.

### List of changes proposed in this PR (pull-request)

<!-- We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

* New file: zero-to-binder-julia.md


### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

- [ ] Everything looks ok?


### Acknowledging contributors

<!-- Please select the correct box -->

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file: <!-- replace this text with the GitHub IDs of any new contributors -->

sgibson91 commented:
@all-contributors please add @ornithos for tutorial

sgibson91 commented:
@all-contributors please add @ots22 for tutorial

sgibson91 commented:
Hi @davidanthoff, if you have time to quickly review this I'd really appreciate it!

ornithos commented on workshops/boost-research-reproducibility-binder/workshop-presentations/zero-to-binder-julia.md:
````diff
@@ -0,0 +1,296 @@
+
+Now visualise the data by creating a new notebook ("New" :arrow_right: "Julia") and run the following code in a cell.
+
+```julia
+using DataFrames
+using CSV
+using Plots
+
+data = CSV.read("gapminder.csv")
+
+# Transpose DataFrame
+numeric_tr = Matrix(data[:,2:end])'
+data_tr = DataFrame(numeric_tr, Symbol.(data[:, :country]))
+
+# Extract Year and add to DataFrame
+column_names = names(data)
+str_years = [String(x)[end-3:end] for x in column_names[2:end]]
+years = parse.(Int, str_years)
+data_tr[:, :Year] = years
+
+# Plot
+plot(data_tr[:, :Year], data_tr[:, :Australia])
+```
````
I know this is my own code, but it was developed to mimic the actions of the Python variant. Since it doesn't use built-in `pandas` functions, it's much less clear why we might want to do this or indeed what it's doing. Even as a Julia user, I don't find it especially intuitive.

One might consider something simpler if the plot is the primary purpose of this:

```julia
using DataFrames
using CSV
using Plots

data = CSV.read("gapminder.csv")

# Extract the row corresponding to Australia
aus_gdp = data[data[:, :country] .== "Australia", :]
aus_gdp = Matrix(aus_gdp[:,2:end])[:]  # as vector

# Extract the years as Ints from the column names
years = [x[end-3:end] for x in names(data)[2:end]]
years = parse.(Int, years)

# Plot
plot(years, aus_gdp)
```

sgibson91 replied:
Yes, thank you. The plot is the purpose of this section and a crazy dream of mine is that I could run a Zero to Binder tutorial and attendees could choose their preferred language within the same session. The point of the tutorial is to learn how to use Binder for Julia users, not get caught up with the workings of Julia. So I agree, let's go with the more intuitive example.

sgibson91 commented on workshops/boost-research-reproducibility-binder/workshop-presentations/zero-to-binder-julia.md:
```diff
@@ -0,0 +1,296 @@
+
+Based on Tim Head's _Zero-to-Binder_ workshops which can be found here: <http://bit.ly/zero-to-binder> and <http://bit.ly/zero-to-binder-rise>
+
+To follow these instructions on your own machine, follow this link: **<http://bit.ly/zero-to-binder-julia>**
```
That's because 1) the bitly link doesn't exist yet, and 2) when it does, it'll point to the default branch. This way, the link remains consistent even if I update the tutorial. Thanks for checking though :)

sgibson91 replied:
That's because 1) the bitly link doesn't exist yet, and 2) when it does, it'll point to the default branch. This way, the link remains consistent even if I update the tutorial. Thanks for checking though :)

sgibson91 commented:
Thank you for your review Alex!

> My main comment is how painfully slow all of this with.

Fun fact: R is also slow in Binder because there are no binaries available for linux (which is what's running on Binder's machines) so we have to compile everything from source.

I actually like the slowness of the tutorial in this sense as it gives attendees time to catch up, helpers to trouble-shoot, and the session leader to explain what is happening. I totally take your point that this is not what Julia was designed for, and indeed we're not using Binder entirely as it was designed for in this tutorial either (that's partly so the only prerequisite of the tutorial is a GitHub account).

Hopefully, that addresses some of your concerns but also happy to discuss it further?

ornithos commented:
That totally makes sense in the context of a class tutorial. 

Re R -- maybe you should have a race to see if Julia or R builds first to decide which language is better. Finally we have an objective means of resolving a longstanding question.

sgibson91 commented:
@all-contributors please add @ornithos for review

sgibson91 commented on workshops/boost-research-reproducibility-binder/workshop-presentations/zero-to-binder-julia.md:
````diff
@@ -0,0 +1,296 @@
+Now visualise the data by creating a new notebook ("New" :arrow_right: "Julia") and run the following code in a cell.
+
+```julia
+using DataFrames
+using CSV
+using Plots
+
+data = CSV.read("gapminder.csv")
+
+# Transpose DataFrame
+numeric_tr = Matrix(data[:,2:end])'
+data_tr = DataFrame(numeric_tr, Symbol.(data[:, :country]))
+
+# Extract Year and add to DataFrame
+column_names = names(data)
+str_years = [String(x)[end-3:end] for x in column_names[2:end]]
+years = parse.(Int, str_years)
+data_tr[:, :Year] = years
+
+# Plot
+plot(data_tr[:, :Year], data_tr[:, :Australia])
````
```suggestion
using DataFrames
using CSV
using Plots

data = CSV.read("gapminder.csv")

# Extract the row corresponding to Australia
aus_gdp = data[data[:, :country] .== "Australia", :]
aus_gdp = Matrix(aus_gdp[:,2:end])[:]  # as vector

# Extract the years as Ints from the column names
years = [x[end-3:end] for x in names(data)[2:end]]
years = parse.(Int, years)

# Plot
plot(years, aus_gdp)
```
